server/service/schedule/schedule.go

```diff
@@ -139,126 +155,206 @@ func New(
-					continue
-				}
-				if newInterval <= 0 {
-					level.Debug(s.logger).Log("msg", "config reload interval method returned invalid interval")
-					continue
+					ctxerr.Handle(s.ctx, err)
 				}
-				if schedInterval == newInterval {
-					level.Debug(s.logger).Log("msg", "schedule interval unchanged")
-					continue
+
+				for _, job := range s.jobs {
+					level.Debug(s.logger).Log("msg", "starting", "jobID", job.ID)
+					if err := runJob(s.ctx, job.Fn); err != nil {
+						level.Error(s.logger).Log("err", "running job", "details", err, "jobID", job.ID)
+						sentry.CaptureException(err)
+						ctxerr.Handle(s.ctx, err)
+					}
 				}
-				s.setSchedInterval(newInterval)
+				level.Info(s.logger).Log("status", "completed")
 
-				newWait := 10 * time.Millisecond
-				if time.Since(currStart) < newInterval {
-					newWait = newInterval - time.Since(currStart)
+				if err := s.updateStats(statsID, fleet.CronStatsStatusCompleted); err != nil {
+					level.Error(s.logger).Log("err", fmt.Sprintf("update cron stats %s", s.name), "details", err)
+					sentry.CaptureException(err)
+					ctxerr.Handle(s.ctx, err)
 				}
-				setWaitTimes(currStart, newWait)
-				schedTicker.Reset(newWait)
 
-				level.Debug(s.logger).Log("new schedule interval", newInterval, "new wait", newWait)
+				// we need to re-synchronize this schedule instance so that the next scheduled run
+				// starts at the beginning of the next full interval
+				//
+				// for example, if the interval is 1hr and the schedule takes 0.2 hrs to run
+				// then we wait 0.8 hrs until the next time we run the schedule, or if the
+				// the schedule takes 1.5 hrs to run then we wait 0.5 hrs (skipping the scheduled
+				// tick that would have overlapped with the 1.5hrs running time)
+				schedInterval = s.getSchedInterval()
+				if time.Since(newStart) > schedInterval {
+					level.Info(s.logger).Log("msg", fmt.Sprintf("total runtime (%v) exceeded schedule interval (%v)", time.Since(newStart), schedInterval))
+					newStart = newStart.Add(time.Since(newStart).Truncate(schedInterval)) // advances start time by the number of full interval elasped
+				}
+				remainingInterval = getRemainingInterval(newStart, schedInterval)
+				clearTickerChannel(schedTicker) // in case another tick arrived during this run
+				schedTicker.Reset(remainingInterval)
+				cancelHold()
 			}
 		}
 	}()
 
+	if s.configReloadIntervalFn != nil {
+		// WithConfigReloadInterval option applies so we periodically check for config updates and
+		// reset the schedInterval for the previous loop
+		g.Add(+1)
+		go func() {
+			defer g.Done()
+
+			configTicker := time.NewTicker(s.configReloadInterval)
+			for {
+				select {
+				case <-s.ctx.Done():
+					configTicker.Stop()
+					return
+				case <-configTicker.C:
+					schedInterval := s.getSchedInterval()
+					newInterval, err := s.configReloadIntervalFn(s.ctx)
+					newInterval = truncateSecondsWithFloor(newInterval)
+					if err != nil {
```

Comment:
In general, you want to check for `err != nil` _before_ using a return value from a function, because you can't make any assurances about the value itself. In this case, I'm guessing it returns 0, but for the sake of following patterns, I would switch it around.